### PR TITLE
[IMP] account : remove internal transfers

### PR DESCRIPTION
**Pull request by jbw-odoo**
Needed for payments refactor

Task : 2199740

### CODE REVIEW

File: addons/account/models/account_payment.py [hunk 1]
```diff
@@ -434,201 +422,131 @@ def unlink(self):
...
-                        rec_pay_line_name += _("Vendor Credit Note")
-                    elif payment.payment_type == 'outbound':
-                        rec_pay_line_name += _("Vendor Payment")
-                if payment.invoice_ids:
-                    rec_pay_line_name += ': %s' % ', '.join(payment.invoice_ids.mapped('name'))
-
-            # Compute 'name' to be used in liquidity line.
-            if payment.payment_type == 'transfer':
-                liquidity_line_name = _('Transfer to %s') % payment.destination_journal_id.name
+        self.ensure_one()
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: unnecesarry change, and it fucks up the whole diff :-)

so, keep the `for` loop

File: addons/account/models/account_payment.py [hunk 2]
```diff
@@ -637,7 +555,7 @@ def post(self):
             If invoice_ids is not empty, there will be one reconcilable move line per invoice to reconcile with.
             If the payment is a transfer, a second journal entry is created in the destination journal to receive money from the transfer account.
         """
-        AccountMove = self.env['account.move'].with_context(default_move_type='entry')
+        AccountMove = self.env['account.move'].with_context(default_type='entry')
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: unrelated diff?

Comment (jbw-odoo) [reply addons/account/models/account_payment.py]: appears to be a rebase error

File: addons/account/models/account_payment.py (hunk 1 above)
Comment (jbw-odoo) [reply addons/account/models/account_payment.py]: will do

File: addons/account/models/account_payment.py [hunk 3]
```diff
@@ -70,7 +68,7 @@ class account_payment(models.Model):
 
     partner_type = fields.Selection([('customer', 'Customer'), ('supplier', 'Vendor')], tracking=True, readonly=True, states={'draft': [('readonly', False)]})
     partner_id = fields.Many2one('res.partner', string='Partner', tracking=True, readonly=True, states={'draft': [('readonly', False)]}, domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]")
-
+    payment_is_internal = fields.Boolean(compute="_compute_payment_is_internal")
```
Comment (smetl) [inline addons/account/models/account_payment.py]: why not `is_internal_transfer`?

File: addons/account/models/account_payment.py [hunk 4]
```diff
@@ -345,15 +344,15 @@ def _get_move_reconciled(self):
...
+    @api.depends('invoice_ids', 'payment_type', 'partner_type', 'partner_id', 'journal_id')
     def _compute_destination_account_id(self):
         for payment in self:
             payment = payment.with_company(payment.company_id)
             if payment.invoice_ids:
                 payment.destination_account_id = payment.invoice_ids[0].mapped(
                     'line_ids.account_id').filtered(
                         lambda account: account.user_type_id.type in ('receivable', 'payable'))[0]
-            elif payment.payment_type == 'transfer':
+            elif payment.payment_is_internal or not payment.partner_id:
```
Comment (smetl) [inline addons/account/models/account_payment.py]: I don't get the point of `or not payment.partner_id`

File: addons/account/models/account_payment.py [hunk 5]
```diff
@@ -502,27 +491,23 @@ def _prepare_payment_moves(self):
...
+                elif payment.payment_type == 'outbound':
+                    rec_pay_line_name += _("Vendor Payment")
+            if payment.invoice_ids:
+                rec_pay_line_name += ': %s' % ', '.join(payment.invoice_ids.mapped('name'))
 
             # Compute 'name' to be used in liquidity line.
-            if payment.payment_type == 'transfer':
-                liquidity_line_name = _('Transfer to %s') % payment.destination_journal_id.name
-            else:
-                liquidity_line_name = payment.name
```
Comment (smetl) [inline addons/account/models/account_payment.py]: outbound + internal transfer = `Transfer from <journal>`
inbound + internal transfer = `Transfer to <journal>`

File: addons/account/models/account_payment.py [hunk 6]
```diff
@@ -649,41 +583,32 @@ def post(self):
 
             # keep the name in case of a payment reset to draft
             if not rec.name:
-                # Use the right sequence to set the name
-                if rec.payment_type == 'transfer':
+                if rec.payment_is_internal or not rec.partner_type:
```
Comment (smetl) [inline addons/account/models/account_payment.py]: `or not rec.partner_type` ?

File: addons/account/tests/test_account_payment.py [hunk 7]
```diff
@@ -176,117 +172,3 @@ def test_outbound_payment_move_1(self):
             },
             expected_payable_line,
         ])
-
-    def test_transfer_payment_move_1(self):
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: need new tests for internal transfer

File: addons/account/models/account_journal_dashboard.py [hunk 8]
```diff
@@ -531,10 +526,6 @@ def create_supplier_payment(self):
         """return action to create a supplier payment"""
         return self.open_payments_action('outbound', mode='form')
 
-    def create_internal_transfer(self):
-        """return action to create a internal transfer"""
-        return self.open_payments_action('transfer', mode='form')
```
Comment (smetl) [inline addons/account/models/account_journal_dashboard.py]: Should't we kept this action from the dashboard?

File: addons/account/tests/test_account_payment.py [hunk 9]
```diff
@@ -20,25 +20,22 @@ def setUpClass(cls):
...
         payment_form.journal_id = cls.bank_journal
         payment_form.payment_date = fields.Date.from_string('2019-01-01')
         payment_form.amount = 100
-        if payment_type == 'transfer':
-            payment_form.destination_journal_id = cls.cash_journal
-        else:
-            payment_form.partner_type = partner_type
-            payment_form.partner_id = cls.partner_a
+        payment_form.partner_type = partner_type
+        payment_form.partner_id = cls.partner_a
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: ` = partner_id` ?

File: addons/account/tests/test_account_payment.py [hunk 10]
```diff
@@ -177,116 +174,91 @@ def test_outbound_payment_move_1(self):
...
-        liquidity_bank_account = self.bank_journal.default_debit_account_id
-        liquidity_cash_account = self.cash_journal.default_credit_account_id
+    def _init_internal_transfer(cls, payment_type, journal_id):
+        payment_form = Form(cls.env['account.payment'])
+        payment_form.payment_type = payment_type
+        payment_form.journal_id = journal_id
+        payment_form.payment_date = fields.Date.from_string('2019-01-01')
+        payment_form.amount = 100
+        payment_form.partner_id = cls.env.company.partner_id
+        return payment_form.save()
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: set this init at the beginning of the file plz

Comment (smetl) [reply addons/account/tests/test_account_payment.py]: Also, don't use `env.company` but `cls.company_data['company']`

File: addons/account/tests/test_account_payment.py [hunk 11]
```diff
@@ -177,116 +174,91 @@ def test_outbound_payment_move_1(self):
...
 
-        cash_transfer_line = {
+    def assertPaymentValues(self, payment, expected_lines_values):
+        self.assertRecordValues(payment.move_line_ids.sorted(lambda line: (line.name, line.balance)), expected_lines_values)
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: move back this helper at the top of the file

File: addons/account/tests/test_account_payment.py [hunk 12]
```diff
@@ -20,25 +20,22 @@ def setUpClass(cls):
 
         cls.inbound_payment = cls._init_payment(cls, 'inbound', partner_type='customer')
         cls.outbound_payment = cls._init_payment(cls, 'outbound', partner_type='supplier')
-        cls.transfer_payment = cls._init_payment(cls, 'transfer')
+        cls.bank_transfer_payment_in = cls._init_internal_transfer(cls, 'inbound', cls.bank_journal)
+        cls.bank_transfer_payment_out = cls._init_internal_transfer(cls, 'outbound', cls.bank_journal)
+        cls.cash_transfer_payment_in = cls._init_internal_transfer(cls, 'inbound', cls.cash_journal)
+        cls.cash_transfer_payment_out = cls._init_internal_transfer(cls, 'outbound', cls.cash_journal)
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: testing both cash/bank seems overkill to me. Why only one is not enough?

File: addons/account/tests/test_account_payment.py [hunk 13]
```diff
@@ -177,116 +174,91 @@ def test_outbound_payment_move_1(self):
...
+
+    def test_transfer_payment_1(self):
+        liquidity_bank_account = self.bank_journal.default_debit_account_id
+        liquidity_cash_account = self.cash_journal.default_credit_account_id  ####
+        self.bank_transfer_payment_in.post()
+        self.cash_transfer_payment_in.post()
+        transfer_line = {
             'account_id': self.company_data['company'].transfer_account_id.id,
-            'partner_id': False,
+            'partner_id': self.env.company.partner_id.id,
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: `self.company_data['company'].partner_id.id`

File: addons/account/tests/test_account_payment.py [hunk 14]
```diff
@@ -177,116 +174,91 @@ def test_outbound_payment_move_1(self):
...
             'amount_currency': 0.0,
             'debit': 0.0,
             'credit': 100.0,
         }
-        bank_transfer_line = {
-            'account_id': self.company_data['company'].transfer_account_id.id,
-            'partner_id': False,
+        liquidity_bank_line = {
+            'account_id': self.bank_journal.default_credit_account_id.id,
+            'partner_id': self.env.company.partner_id.id,
```
Comment (smetl) [inline addons/account/tests/test_account_payment.py]: `self.company_data['company'].partner_id.id`

File: addons/account/models/account_journal_dashboard.py [hunk 15]
```diff
@@ -493,6 +493,8 @@ def open_payments_action(self, payment_type, mode='tree'):
             action_ref = 'account.action_account_payments'
         [action] = self.env.ref(action_ref).read()
         action['context'] = dict(ast.literal_eval(action.get('context')), default_journal_id=self.id, search_default_journal_id=self.id)
+        if payment_type == 'transfer':
```
Comment (qdp-odoo) [inline addons/account/models/account_journal_dashboard.py]: what's that?

Comment (smetl) [reply addons/account/models/account_journal_dashboard.py]: Called by the dashboard here:
https://github.com/odoo/odoo/blob/master/addons/account/views/account_journal_dashboard_view.xml#L71

This will not be set on an account.payment :)

**Comment by smetl** [comment]
@robodoo r+
